### PR TITLE
Fix bug in marshaling/unmarshaling Time in protobufs

### LINKED ISSUE #25347: Deployment tests are panic-ing when run with protobufs
Example stack:

      Test Panicked
      runtime error: invalid memory address or nil pointer dereference
      /usr/local/go/src/runtime/panic.go:443

      Full Stack Trace
      	/usr/local/go/src/runtime/panic.go:443 +0x4e9
      k8s.io/kubernetes/test/e2e/framework.WaitForDeploymentRevisionAndImage.func1(0x1991f00, 0x0, 0x0)
  	    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2585 +0x2f2
      k8s.io/kubernetes/pkg/util/wait.WaitFor(0xc820ac37e0, 0xc820973c00, 0xc8203c0cc0, 0x0, 0x0)
      	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:181 +0x77
      k8s.io/kubernetes/pkg/util/wait.pollInternal(0xc820ac37e0, 0xc820973c00, 0x0, 0x0)
      	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:143 +0x8a
      k8s.io/kubernetes/pkg/util/wait.Poll(0x77359400, 0xdf8475800, 0xc820973c00, 0x0, 0x0)
      	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:137 +0x4f
      k8s.io/kubernetes/test/e2e/framework.WaitForDeploymentRevisionAndImage(0x7f2474238528, 0xc82069d020, 0xc8207142a0, 0x1a, 0x2371ba0, 0x18, 0x2195280, 0x1, 0x23dd2c0, 0x22, ...)
  	    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2572 +0x19c
      k8s.io/kubernetes/test/e2e.testRecreateDeployment(0xc8200fedc0)
  	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:390 +0xf8a
      k8s.io/kubernetes/test/e2e.glob.func14.4()
      	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:63 +0x20
      k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc820427920, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
      	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:230 +0x42c
      k8s.io/kubernetes/test/e2e.TestE2E(0xc8200c5cb0)
      	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:30 +0x21
      testing.tRunner(0xc8200c5cb0, 0x328b820)
      	/usr/local/go/src/testing/testing.go:473 +0x98
      created by testing.RunTests
      	/usr/local/go/src/testing/testing.go:582 +0x892

@smarterclayton

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix #25347

### CODE REVIEW

**Comment by wojtek-t** [comment]
Unfortunately the fix doesn't work - e2e cluster with protobufs is not starting with the following failure:

    F0510 14:31:16.906620       8 controller.go:86] Unable to perform initial IP allocation check: unable to refresh the service IP block: provided data does not appear to be a protobuf message, expected prefix [107 56 115 0]

**Comment by smarterclayton** [comment]
That implies that the server is not starting with the correct data.   It's
possible that this got broken by Nikhil's recent change to genericapiserver
(I was having to rebase today).

**Comment by wojtek-t** [comment]
@smarterclayton - I'm pretty sure it's realted to my change, because we have e2e tests with protobufs running continuously from head, and the cluster is starting correctly there.

**Comment by smarterclayton** [comment]
Ok

File: pkg/api/unversioned/time_proto.go [hunk 1]
```diff
@@ -51,20 +52,30 @@ func (m *Time) Size() (n int) { return m.ProtoTime().Size() }
...
-	if err := p.Unmarshal(data); err != nil {
-		return err
+	if len(data) == 0 {
+		m.Time = time.Time{}
+	} else {
+		p := Timestamp{}
+		if err := p.Unmarshal(data); err != nil {
+			return err
+		}
+		m.Time = time.Unix(p.Seconds, int64(p.Nanos)).Local()
```
Comment (smarterclayton) [inline pkg/api/unversioned/time_proto.go]: Pretty sure this needs to be UTC().  Will double check.

Comment (smarterclayton) [inline pkg/api/unversioned/time_proto.go]: Nm, I saw we were using local internally for the other types.

**Comment by smarterclayton** [comment]
Found the issue, commented in the other with a diff of what needs to be changed.

**Comment by wojtek-t** [comment]
I'm reassigning for @gmarek to have it in ASAP.
@smarterclayton - since you already saw this (and send me diff) I think you're fine with that. If not, I will apply your comments in a separate PR.

**Comment by wojtek-t** [comment]
Manually self-merging to see what other problems do we have with protobufs.

**Comment by wojtek-t** [comment]
Ref #8132

**Pull request merged**